### PR TITLE
Use custom QueryPagingSource

### PR DESCRIPTION
**Pull request by ghostbear**
- Adds placeholders to make the list jump around less
- Fixes the issue where SQLDelights QueryPagingSource would throw IndexOutOfBounds

https://user-images.githubusercontent.com/6576096/174432964-a0145fac-6339-4158-b633-78e2d568196a.mp4

### CODE REVIEW

File: app/src/main/java/eu/kanade/presentation/history/components/HistoryItem.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            Icon(
+                imageVector = Icons.Outlined.Delete,
+                contentDescription = stringResource(R.string.action_delete),
+                tint = MaterialTheme.colorScheme.onSurface,
+            )
+        }
+    }
+}
+
+@Composable
```
Comment (arkon) [inline app/src/main/java/eu/kanade/presentation/history/components/HistoryItem.kt:97]: FWIW: https://google.github.io/accompanist/placeholder/#shimmer

File: app/src/main/java/eu/kanade/presentation/history/HistoryScreen.kt [hunk 2]
```diff
@@ -107,6 +98,34 @@ fun HistoryContent(
     var removeState by remember { mutableStateOf<HistoryWithRelations?>(null) }
 
     val scrollState = rememberLazyListState()
+
+    val gradient = listOf(
```
Comment (arkon) [inline app/src/main/java/eu/kanade/presentation/history/HistoryScreen.kt:102]: Probably want to pull these out to somewhere generic for reuse if we keep these.

File: app/src/main/java/eu/kanade/data/QueryPagingSource.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    override fun getRefreshKey(state: PagingState<Long, RowType>): Long? {
+        return state.anchorPosition?.let { anchorPosition ->
+            val anchorPage = state.closestPageToPosition(anchorPosition)
+            anchorPage?.prevKey ?: anchorPage?.nextKey
+        }
+    }
+
+    override fun queryResultsChanged() {
+        invalidate()
```
Comment (arkon) [inline app/src/main/java/eu/kanade/data/QueryPagingSource.kt:72]: Is it necessary to invalidate everything? Does this trigger even if it's a single item being changed (state, not position)?

File: app/src/main/java/eu/kanade/presentation/history/components/HistoryItem.kt (hunk 1 above)
Comment (ghostbear) [reply app/src/main/java/eu/kanade/presentation/history/components/HistoryItem.kt:97]: I feel like doing it manually is fine

File: app/src/main/java/eu/kanade/data/QueryPagingSource.kt (hunk 3 above)
Comment (ghostbear) [reply app/src/main/java/eu/kanade/data/QueryPagingSource.kt:70]: Yes as some things may be on a page edge. It only triggers on database changes so if an item is inserted, updated deleted, etc

**Pull request merged**